### PR TITLE
Integration test improvements

### PR DESCRIPTION
**Pull request by ArcSin2000X**
Folders for test data are now created with unique names to prevent problems with delete. Improvements to  TestHelper to preven locking main thread. Increased default time for WaitLoop freaquency

### CODE REVIEW

File: src/Stratis.Bitcoin.Tests.Common/TestBase.cs [hunk 1]
```diff
@@ -25,31 +25,10 @@ public TestBase(Network network)
...
-                else
-                    break;
-            }
-
-            if (deleteAttempts >= 50)
-                throw new Exception(string.Format("The test folder: {0} could not be deleted.", dir));
-
-            Directory.CreateDirectory(dir);
-            return dir;
+            long ticks = DateTime.UtcNow.Ticks;
```
Comment (zeptin) [inline src/Stratis.Bitcoin.Tests.Common/TestBase.cs]: This seems like it is going to get really messy; how do you suggest we ensure the created directories don't accumulate forever? On the CI I assume the folders for each test run are isolated from each other regardless, so this is mostly for local tests?

File: src/Stratis.Bitcoin.IntegrationTests.Common/TestHelper.cs [hunk 2]
```diff
@@ -11,17 +12,23 @@ namespace Stratis.Bitcoin.IntegrationTests.Common
...
+        static async Task AsyncTaskDelay(int milisec)
+        {
+            await Task.Delay(milisec).ConfigureAwait(false);
+        }
+
+        public static void WaitLoop(Func<bool> act, string failureReason = "Unknown Reason", int millisecondsTimeout = 150, CancellationToken cancellationToken = default(CancellationToken))
         {
             cancellationToken = cancellationToken == default(CancellationToken)
-                ? new CancellationTokenSource(Debugger.IsAttached ? 15 * 60 * 1000 : 30 * 1000).Token
+                ? new CancellationTokenSource(Debugger.IsAttached ? 15 * 60 * 1000 : 40 * 1000).Token
```
Comment (zeptin) [inline src/Stratis.Bitcoin.IntegrationTests.Common/TestHelper.cs]: Can you give a bit of detail as to how you arrived at this particular choice of timeout? I also find that tests seem to be more reliable the longer this interval is, but perhaps the underlying tests should be corrected instead? Do we know which ones are particularly problematic?

File: src/Stratis.Bitcoin.Tests.Common/TestBase.cs [hunk 3]
```diff
@@ -25,31 +25,10 @@ public TestBase(Network network)
...
+            string uniqueDirName = dir + ticks.ToString();
```
Comment (bokobza) [inline src/Stratis.Bitcoin.Tests.Common/TestBase.cs]: ticks is not very readable. It'd be hard to quickly see the results of the tests you ran a couple of days ago.
I prefer `$"{dir}-{DateTime.UtcNow.ToString("dd/MM/yyyyTHH:mm:ss.fff")}"` which will give you folders like _forldername_-15/08/2018T15:40:43.144

File: src/Stratis.Bitcoin.IntegrationTests.Common/TestHelper.cs [hunk 4]
```diff
@@ -11,17 +12,23 @@ namespace Stratis.Bitcoin.IntegrationTests.Common
...
                 : cancellationToken;
+
             while (!act())
             {
                 try
                 {
                     cancellationToken.ThrowIfCancellationRequested();
-                    Thread.Sleep(millisecondsTimeout);
+                    AsyncTaskDelay(millisecondsTimeout).Wait();
```
Comment (bokobza) [inline src/Stratis.Bitcoin.IntegrationTests.Common/TestHelper.cs]: So I think the `Wait()` here defeats the purpose of having a `Task.Delay` as the task is waited on. 
Running the code, I saw no difference in speed or reliability.

File: src/Stratis.Bitcoin.IntegrationTests.Common/TestHelper.cs [hunk 5]
```diff
@@ -11,17 +12,23 @@ namespace Stratis.Bitcoin.IntegrationTests.Common
 {
     public class TestHelper
     {
-        public static void WaitLoop(Func<bool> act, string failureReason = "Unknown Reason", int millisecondsTimeout = 50, CancellationToken cancellationToken = default(CancellationToken))
+        static async Task AsyncTaskDelay(int milisec)
```
Comment (ferdeen) [inline src/Stratis.Bitcoin.IntegrationTests.Common/TestHelper.cs]: `millisecs` ?

File: src/Stratis.Bitcoin.Tests.Common/TestBase.cs (hunk 1 above)
Comment (ArcSin2000X) [reply src/Stratis.Bitcoin.Tests.Common/TestBase.cs]: 2 use cases:
1) you run tests on server after build, according to @bokobza you always start with empty dir, that way its not messy
2) you run locally, we will need to housekeep directories, I found a lot of problems when run locally and trying to delete dirs from C# this way you are guaranteed new folder every time

File: src/Stratis.Bitcoin.IntegrationTests.Common/TestHelper.cs (hunk 4 above)
Comment (bokobza) [reply src/Stratis.Bitcoin.IntegrationTests.Common/TestHelper.cs]: https://stackoverflow.com/questions/34052381/thread-sleep2500-vs-task-delay2500-wait

File: src/Stratis.Bitcoin.Tests.Common/TestBase.cs (hunk 3 above)
Comment (zeptin) [reply src/Stratis.Bitcoin.Tests.Common/TestBase.cs]: +1, but you may need to use underscores etc in lieu of characters that are invalid in folder names.

File: src/Stratis.Bitcoin.IntegrationTests.Common/TestHelper.cs (hunk 2 above)
Comment (ArcSin2000X) [reply src/Stratis.Bitcoin.IntegrationTests.Common/TestHelper.cs]: we managed to isolate around 12 tests that cause problems if executed in integration test suite. I think given large peformance profile of the test its just giving extra 10 seconds for the loop to spin, additionally I decreased the frequency at which we invoke methods

File: src/Stratis.Bitcoin.Tests.Common/TestBase.cs (hunk 3 above)
Comment (ArcSin2000X) [reply src/Stratis.Bitcoin.Tests.Common/TestBase.cs]: accept

File: src/Stratis.Bitcoin.IntegrationTests.Common/TestHelper.cs (hunk 4 above)
Comment (ArcSin2000X) [reply src/Stratis.Bitcoin.IntegrationTests.Common/TestHelper.cs]: accept

File: src/Stratis.Bitcoin.IntegrationTests.Common/TestHelper.cs (hunk 5 above)
Comment (ArcSin2000X) [reply src/Stratis.Bitcoin.IntegrationTests.Common/TestHelper.cs]: I stand corrected

**Comment by codingupastorm** [comment]
Big fan of this. 

I think we should create an issue to address the underlying problems with tests failing in the current times - some of them are smart contract ones (sorry), but for an interim solution I like the idea of this.

File: src/Stratis.Bitcoin.Tests.Common/TestBase.cs [hunk 6]
```diff
@@ -25,31 +25,9 @@ public TestBase(Network network)
 
         public static string AssureEmptyDir(string dir)
         {
-            int deleteAttempts = 0;
```
Comment (monsieurleberre) [inline src/Stratis.Bitcoin.Tests.Common/TestBase.cs]: Would it be possible / desirable to clean folders from previous runs ? I can help you write the code if the is something we want to do 👍

**Pull request merged**